docs/blueprints/forgotten-deletes.md

```diff
@@ -0,0 +1,118 @@
+
+```
+message DeletePiecesRequest {
+    repeated bytes piece_ids = 1 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+    bytes satellite_signature = 2;
+}
+```
+
+The storage nodes will serialize and save these signed messages into a new database table.
+
+```sql
+CREATE TABLE delete_pieces_requests (
+    id INTEGER PRIMARY KEY,
+    data BYTEA,
+    created_at TIMESTAMP,
+)
+```
+
+The storage node now has proof that the satellite requested these pieces to be deleted. Now we need the ability to access this signed message if given a particular piece ID.
+
+These signed messages may contain multiple piece IDs. We need to be able to reference this row if given any one piece ID contained within the message.
+
+To do this we create another table to hold each deleted piece with a foreign key which references the corresponding row in delete_pieces_requests to retrieve the signed delete request.
+
+```sql
+CREATE TABLE deleted_pieces (
+    id BYTEA,
+    signed_message INTEGER,
+    FOREIGN KEY(signed_message) REFERENCES deleted_pieces_requests(id)
+)
+```
+
+To avoid failing the audit, we want the node to be able to provide proof of the delete request to the satellite right away. To do this we will add the signed DeletePiecesRequest to the PieceDownloadResponse protobuf.
+
+```
+message PieceDownloadResponse {
+    // Chunk response for download request
+    message Chunk {
+        int64 offset = 1;
+        bytes data = 2;
+    }
+    Chunk chunk = 1;
+    orders.PieceHash hash = 2;
+    orders.OrderLimit limit = 3;
+    
+    DeletePiecesRequest delete_pieces_request = 4;
+}
+```
+
+When the satellite receives a PieceDownloadResponse and DeletePiecesRequest is not nil, it will verify the signature and check for the corresponding piece ID. If these checks pass, the audit will not count as a failure. Further, upon receipt of the first verified DeletePiecesRequest, this means that the entire object to which this segment belongs should not exist and we should delete it.
+
+### Garbage collection
+
+Garbage collection works by observing the metainfo loop. This collects all the piece IDs which a node should have according to the pointer DB at that time. The satellite then sends these piece IDs to the storage node, indicating that it should keep these pieces and all pieces which were created after the creation of the bloom filter and delete the rest.
+
+Instead of proving the pieces were deleted, we will remove the possibility of these pieces being restored.
+To acheive this we will simply run garbage collection from a DB snapshot at a point in time beyond which we will never restore. That is, we must decide what is the furthest possible point the past we may restore to and only run garbage collection on a snapshot at or before this point. If this is guaranteed, no pieces deleted by garbage collection should ever be reverted by a DB restoration.
```

Comment:
oh, this is so smart, i love this

Reply:
Gotta give credit to Jeff on this!

Reply:
This time can also be used to determine how long we need to keep track of piece deletion requests for.

Reply:
How would this be accomplished? We will need to decouple garbage collection from the metainfo loop, because I think other metainfo loop observers need current data. Does this add an operations consideration for db snapshots to be served as live databases and for the satellite to have read access to them? Or perhaps the gc metainfo loop would just read a db snapshot in pgdump format without needing SQL? Or does this mean we would use the cockroach "AS OF SYSTEM TIME" syntax and make postgresql no longer supported?